### PR TITLE
[freeboxos] Add websocket connection refresh mechanism

### LINKED ISSUE #15534: [freeboxos] Access to freebox server goes to "Erreur interne"
After some time (~ 24h), I observed that the freebox server goes on error. The web interface displays "Erreur interne" and most of the things goes offline.

Note that everything works fine (network, wifi, tv...)

I suspect this to be related to webhook subscription.

## Expected Behavior
The binding shoud not push the box in error.

## Possible Solution
Brutal : disable webhook subscription
Better : see how connection to the webhook could be opened/closed to avoid this.

## Steps to Reproduce (for Bugs)
Define your bridge, things, especially some host and wifihosts things.
Wait some time (~24h). Things should go offline and connection via web browser display an error message.

### PR DESCRIPTION
**Pull request by clinique**
Resolves #15534.
Unified subscription mechanism between host and vm.

Added a reconnect interval to the API bridge.

### CODE REVIEW

File: bundles/org.openhab.binding.freeboxos/README.md [hunk 1]
```diff
@@ -54,14 +54,15 @@ FreeboxOS binding has the following configuration parameters:
...
+| Freebox Server Address        | apiDomain           | The domain to use in place of hardcoded Freebox ip            | No       | mafreebox.freebox.fr |
+| Application Token             | appToken            | Token generated by the Freebox Server.                        | Yes      |                      |
+| Network Device Discovery      | discoverNetDevice   | Enable the discovery of network device things.                | No       | false                |
+| Background Discovery Interval | discoveryInterval   | Interval in minutes - 0 disables background discovery         | No       | 10                   |
+| HTTPS Available               | httpsAvailable      | Tells if https has been configured on the Freebox             | No       | false                |
+| HTTPS port                    | httpsPort           | Port to use for remote https access to the Freebox Api        | No       | 15682                |
+| Websocket Reconnect Interval  | wsReconnectInterval | Disconnection interval frequency - 0 disables websocket usage | No       | 60                   |
```
Comment (lolodomo) [inline bundles/org.openhab.binding.freeboxos/README.md:65]: Please mention the unit in the description, it is not obvious that this is a value in minutes.

File: bundles/org.openhab.binding.freeboxos/src/main/resources/OH-INF/config/bridge-config.xml [hunk 2]
```diff
@@ -42,6 +42,12 @@
 			<advanced>true</advanced>
 			<default>15682</default>
 		</parameter>
+		<parameter name="wsReconnectInterval" type="integer" min="0" max="1440" required="false" unit="min">
+			<label>Websocket Reconnect Interval</label>
+			<description>Disconnection interval frequency - 0 disables websocket usage</description>
```
Comment (lolodomo) [inline bundles/org.openhab.binding.freeboxos/src/main/resources/OH-INF/config/bridge-config.xml:47]: Please mention the unit in the description, it is not obvious that this is a value in minutes.

File: bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/rest/WebSocketManager.java [hunk 3]
```diff
@@ -82,25 +91,49 @@ public String getEvent() {
...
+        if (reconnectInterval > 0) {
+            WebSocketClient client = new WebSocketClient(apiHandler.getHttpClient());
+            URI uri = getUriBuilder().scheme(getUriBuilder().build().getScheme().contains("s") ? "wss" : "ws").build();
+            ClientUpgradeRequest request = new ClientUpgradeRequest();
+            request.setHeader(ApiHandler.AUTH_HEADER, sessionToken);
+
+            stopReconnect();
+            ScheduledExecutorService scheduler = ThreadPoolManager
+                    .getScheduledPool(FreeboxOsBindingConstants.BINDING_ID);
+            reconnectJob = Optional.of(scheduler.scheduleAtFixedRate(() -> {
```
Comment (lolodomo) [inline bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/rest/WebSocketManager.java:104]: https://www.openhab.org/docs/developer/guidelines.html#e-runtime-behavior
Recommendation is to prefer scheduleWithFixedDelay over scheduleAtFixedRate.
Do you really need scheduleAtFixedRate ?

File: bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/rest/WebSocketManager.java [hunk 4]
```diff
@@ -82,25 +91,49 @@ public String getEvent() {
...
+        reconnectJob.ifPresent(job -> job.cancel(true));
+        reconnectJob = Optional.empty();
+    }
+
+    public void dispose() {
+        stopReconnect();
+        closeSession();
+    }
+
+    private void closeSession() {
```
Comment (lolodomo) [inline bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/rest/WebSocketManager.java:131]: Is it normal that there is no call to `client.stop()` anywhere ?

File: bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/rest/WebSocketManager.java [hunk 5]
```diff
@@ -82,25 +91,49 @@ public String getEvent() {
...
+    public void openSession(@Nullable String sessionToken, int reconnectInterval) {
+        if (reconnectInterval > 0) {
+            WebSocketClient client = new WebSocketClient(apiHandler.getHttpClient());
+            URI uri = getUriBuilder().scheme(getUriBuilder().build().getScheme().contains("s") ? "wss" : "ws").build();
+            ClientUpgradeRequest request = new ClientUpgradeRequest();
+            request.setHeader(ApiHandler.AUTH_HEADER, sessionToken);
+
+            stopReconnect();
+            ScheduledExecutorService scheduler = ThreadPoolManager
+                    .getScheduledPool(FreeboxOsBindingConstants.BINDING_ID);
```
Comment (lolodomo) [inline bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/rest/WebSocketManager.java:103]: How many thread pools will you create by doing that ?
Is it just one for all the binding ?

File: bundles/org.openhab.binding.freeboxos/src/main/resources/OH-INF/i18n/freeboxos.properties [hunk 6]
```diff
@@ -353,6 +359,11 @@ channel-type.freeboxos.wifi-host.label = Access Point
 channel-type.freeboxos.wifi-status.label = Wifi Enabled
 channel-type.freeboxos.wifi-status.description = Indicates whether the wifi network is enabled
 
+# thing types config
+
+thing-type.config.freeboxos.host.mDNS.label = mDNS Name
+thing-type.config.freeboxos.host.mDNS.description = The mDNS name of the network device
+
```
Comment (lolodomo) [inline bundles/org.openhab.binding.freeboxos/src/main/resources/OH-INF/i18n/freeboxos.properties:366]: To be confirmed but I have the feeling that this config parameter does no more exist ?

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.freeboxos/README.md (hunk 1 above)
Comment (clinique) [reply bundles/org.openhab.binding.freeboxos/README.md:65]: Done

File: bundles/org.openhab.binding.freeboxos/src/main/resources/OH-INF/config/bridge-config.xml (hunk 2 above)
Comment (clinique) [reply bundles/org.openhab.binding.freeboxos/src/main/resources/OH-INF/config/bridge-config.xml:47]: Done

File: bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/rest/WebSocketManager.java (hunk 3 above)
Comment (clinique) [reply bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/rest/WebSocketManager.java:104]: Changed

File: bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/rest/WebSocketManager.java (hunk 5 above)
Comment (clinique) [reply bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/rest/WebSocketManager.java:103]: Changed to final member of the class.

File: bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/rest/WebSocketManager.java (hunk 4 above)
Comment (clinique) [reply bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/rest/WebSocketManager.java:131]: Correct, forgot it.

File: bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/rest/WebSocketManager.java [hunk 7]
```diff
@@ -82,25 +95,50 @@ public String getEvent() {
...
+        reconnectJob.ifPresent(job -> job.cancel(true));
+        reconnectJob = Optional.empty();
+    }
 
+    public void dispose() {
+        stopReconnect();
         try {
-            client.start();
-            client.connect(this, uri, request);
+            client.stop();
```
Comment (lolodomo) [inline bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/rest/WebSocketManager.java:129]: Looks inconsistent, you will have more starts (every X minutes) than stops.
Either the stop or the start is not at the right place.

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/rest/WebSocketManager.java (hunk 7 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/rest/WebSocketManager.java:129]: @clinique : did you see my comment?

**Review by clinique — COMMENTED** [review]
@lolodomo : Saw your comment but I did not understand it

File: bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/rest/WebSocketManager.java (hunk 7 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/rest/WebSocketManager.java:129]: What I mean is that you have a threadscheduled every X that start the client. But before you start it, you are not stopping it.

Comment (clinique) [reply bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/api/rest/WebSocketManager.java:134]: Done

**Review by lolodomo — APPROVED** [review]
LGTM, thank you

**Pull request merged**